### PR TITLE
Feature/disable-offer-visibility

### LINKED ISSUE #218: Disable Offer Visibility Buttons While Fulfilling The Request
Disable the visibility buttons on the offer and the company offers management pages while the respective request is being fulfilled. To test in the Chrome dev tools, go to the Network tab, then choose the throttling option "Slow 3G".

### PR DESCRIPTION
**Pull request by Naapperas**
Closes #218

### CODE REVIEW

**Review by TiagooGomess — COMMENTED** [review]
Change the name of the state for loadingOfferVisibility

**Review by TiagooGomess — CHANGES_REQUESTED** [review]
I suggest making this change so that the color of the visibility button when the company or offer are blocked also changes:

image

**Review by TiagooGomess — COMMENTED** [review]
On the offer page, the visibility button does not become disabled while the request is being fulfilled. Can you replicate that on your machine?

**Comment by Naapperas** [comment]
While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hid/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.

**Comment by TiagooGomess** [comment]
> While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hide/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.

You should enable throttling on the dev tools network tab (for example, slow 3G) so that you can simulate a request that takes some time to be fulfilled. Doing that, can you see the visibility button being disabled?

**Comment by TiagooGomess** [comment]
> Also, should i write tests covering these scenarios ?

You should look at the coverage report and see if there are some uncovered lines (that cover added functionality).

**Comment by Naapperas** [comment]
> > While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hide/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.
> 
> You should enable throttling on the dev tools network tab (for example, slow 3G) so that you can simulate a request that takes some time to be fulfilled. Doing that, can you see the visibility button being disabled?

By explicitly waiting 2 seconds (setTimeout) the button appears to be disabled, like this 

image

and it does not fire any events (thus not performing any request to the backend), so we know that the button is disabled at some point. 

What I think is happening is the disabling/enabling is happening faster than the request itself (enabling GPRS or normal 2G, which results in slow requests, didn't seem to make the button get disabled). I added this behavior in 

```js
    const handleOfferVisibility = async () => {
        setLoadingOfferVisibility(true);
        if (visibilityState.isVisible) {
            await handleHideOffer({
                offer: offer,
                addSnackbar: addSnackbar,
                onError: onError,
            });
        } else {
            await handleCompanyEnableOffer({
                offer: offer,
                addSnackbar: addSnackbar,
                onError: onError,
            });
        }
        setLoadingOfferVisibility(false);
    };
```

thinking that the "enabling" part" would only occur after the requests awaited, but it does not seem to be the case. Maybe I could remove the await statement and write this as a then/catch chain, what do you think ?

**Comment by TiagooGomess** [comment]
> > > While testing I left a `console.log` in the OfferVIsibilityOptions file and both the "offer details" on the homepage search section and the offer page itself, when pressing the "hide/enable" button, triggered that log, so I assumed that it worked for both, but i might be mistaken.
> > 
> > 
> > You should enable throttling on the dev tools network tab (for example, slow 3G) so that you can simulate a request that takes some time to be fulfilled. Doing that, can you see the visibility button being disabled?
> 
> By explicitly waiting 2 seconds (setTimeout) the button appears to be disabled, like this
> 
> image
> 
> and it does not fire any events (thus not performing any request to the backend), so we know that the button is disabled at some point.
> 
> What I think is happening is the disabling/enabling is happening faster than the request itself (enabling GPRS or normal 2G, which results in slow requests, didn't seem to make the button get disabled). I added this behavior in
> 
> ```js
>     const handleOfferVisibility = async () => {
>         setLoadingOfferVisibility(true);
>         if (visibilityState.isVisible) {
>             await handleHideOffer({
>                 offer: offer,
>                 addSnackbar: addSnackbar,
>                 onError: onError,
>             });
>         } else {
>             await handleCompanyEnableOffer({
>                 offer: offer,
>                 addSnackbar: addSnackbar,
>                 onError: onError,
>             });
>         }
>         setLoadingOfferVisibility(false);
>     };
> ```
> 
> thinking that the "enabling" part" would only occur after the requests awaited, but it does not seem to be the case. Maybe I could remove the await statement and write this as a then/catch chain, what do you think ?

Yes, you can try that

**Comment by TiagooGomess** [comment]
Not working correctly on my machine:

https://user-images.githubusercontent.com/50210213/162983422-a139dadf-b069-401f-85d0-01cdc3bb1938.mov

**Comment by TiagooGomess** [comment]
When the offer is hidden, the color of the visibility button on the company management offers page is "undefined". It should be secondary. And when the offer is disabled, the color is secondary, but it should be "undefined".

File: src/components/Company/Offers/Manage/CompanyOffersVisibilityActions.js [hunk 1]
```diff
@@ -99,15 +104,30 @@ const CompanyOffersVisibilityActions = ({ offer }) => {
...
     return (
         <Tooltip title={offerVisibilityState.isVisible ? "Hide Offer" : "Enable Offer"}>
-            <IconButton
-                onClick={handleOfferVisibility}
-                disabled={offerVisibilityState.isDisabled || offerVisibilityState.isBlocked}
-            >
-                {offerVisibilityState.isVisible ? <VisibilityOffIcon data-testid="HideOffer" color="secondary" fontSize="medium" />
-                    : <VisibilityIcon data-testid="EnableOffer" color="secondary" fontSize="medium" />}
-            </IconButton>
+            <>
```
Comment (TiagooGomess) [inline src/components/Company/Offers/Manage/CompanyOffersVisibilityActions.js:111]:
```suggestion
            
```

File: src/components/Company/Offers/Manage/CompanyOffersVisibilityActions.js [hunk 2]
```diff
@@ -99,15 +104,30 @@ const CompanyOffersVisibilityActions = ({ offer }) => {
...
+                        />
+                        :
+                        <VisibilityIcon
+                            data-testid="EnableOffer"
+                            color={offerVisibilityButtonDisabled ? undefined : "secondary"}
+                            fontSize="medium"
+                        />
+                    }
+                </IconButton>
+            </>
```
Comment (TiagooGomess) [inline src/components/Company/Offers/Manage/CompanyOffersVisibilityActions.js:130]:
```suggestion
            
```

File: src/components/HomePage/SearchResultsArea/Offer/OfferVisibilityOptions.js [hunk 3]
```diff
@@ -75,31 +73,33 @@ const OfferVisibilityOptions = ({
...
-                                onClick={handleOfferVisibility}
-                                className={classes.visibilityButton}
-                                role="hideEnableOfferButton"
-                                disabled={visibilityState.isDisabled}
-                                startIcon={visibilityState.isVisible ? <VisibilityOff /> : <Visibility />}
-                            >
-                                { visibilityState.isVisible ? "Hide Offer" : "Enable Offer" }
-                            </Button>
+                        <Tooltip title={visibilityState.isVisible ? "Hide Offer" : "Enable Offer"}>
+                            <>
```
Comment (TiagooGomess) [inline src/components/HomePage/SearchResultsArea/Offer/OfferVisibilityOptions.js:80]:
```suggestion
                           
```

File: src/components/HomePage/SearchResultsArea/Offer/OfferVisibilityOptions.js [hunk 4]
```diff
@@ -75,31 +73,33 @@ const OfferVisibilityOptions = ({
...
+                                <Button
+                                    onClick={handleOfferVisibility}
+                                    className={classes.visibilityButton}
+                                    role="hideEnableOfferButton"
+                                    disabled={loadingOfferVisibility || visibilityState.isDisabled}
+                                    startIcon={visibilityState.isVisible ? <VisibilityOff /> : <Visibility />}
+                                >
+                                    {visibilityState.isVisible ? "Hide Offer" : "Enable Offer"}
+                                </Button>
+                            </>
```
Comment (TiagooGomess) [inline src/components/HomePage/SearchResultsArea/Offer/OfferVisibilityOptions.js:90]:
```suggestion
                            
```

**Review by TiagooGomess — CHANGES_REQUESTED** [review]
You must remove these; otherwise, the tooltips don't work.

**Comment by Naapperas** [comment]
> You must remove these; otherwise, the tooltips don't work.

I had changed those because MUI would complain about providing a disabled child to the Tooltip component and I figured that a React fragment would work, changed it to a `span` instead.

**Comment by TiagooGomess** [comment]
Do the same for [OfferVisibilityOptions.js](https://github.com/NIAEFEUP/nijobs-fe/pull/233/files/612dd757c7f466c9640a4a38af9d3e40ed44a67a#diff-5004fe71a0fb5f48ecd3f9cbe5301cce777f84e76a9105d0c0e62ffc81127283)

File: src/services/locationSearchService.js [hunk 5]
```diff
@@ -1,27 +1,37 @@
...
-        return data.map(({ city, country, latitude, longitude }) => ({ city, country, latitude, longitude }));
+        // const data = (await res.json()).data;
+        // return data.map(({ city, country, latitude, longitude }) => ({ city, country, latitude, longitude }));
+
+        const data = await res.json();
+
+        return data;
 
     } catch (err) {
         throw [{ msg: Constants.UNEXPECTED_ERROR_MESSAGE }];
```
Comment (TiagooGomess) [inline src/services/locationSearchService.js:37]: Maybe you should not modify this file in this PR

Comment (Naapperas) [reply src/services/locationSearchService.js:37]: I don't now how this got through...

**Review by TiagooGomess — APPROVED** [review]

**Pull request merged**